crates/loro-internal/src/version/frontiers.rs

```diff
@@ -0,0 +1,582 @@
+            Frontiers::ID(id) => Either::Left(Either::Right(std::iter::once(*id))),
+            Frontiers::Map(map) => Either::Right(map.iter()),
+        }
+    }
+
+    pub fn contains(&self, id: &ID) -> bool {
+        match self {
+            Frontiers::None => false,
+            Frontiers::ID(inner_id) => inner_id == id,
+            Frontiers::Map(map) => map.contains(id),
+        }
+    }
+
+    pub fn push(&mut self, id: ID) {
+        match self {
+            Frontiers::None => *self = Frontiers::ID(id),
+            Frontiers::ID(existing_id) => {
+                if existing_id.peer != id.peer {
+                    let mut map = InternalMap::new();
+                    map.insert(*existing_id);
+                    map.insert(id);
+                    *self = Frontiers::Map(map);
+                } else if existing_id.counter < id.counter {
+                    *existing_id = id;
+                }
+            }
+            Frontiers::Map(map) => map.insert(id),
+        }
+    }
+
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&ID) -> bool,
+    {
+        match self {
+            Frontiers::None => {}
+            Frontiers::ID(id) => {
+                if !f(id) {
+                    *self = Frontiers::None;
+                }
+            }
+            Frontiers::Map(map) => {
+                map.retain(|id| f(id));
+                match map.len() {
+                    0 => *self = Frontiers::None,
+                    1 => {
+                        let id = map.iter().next().unwrap();
+                        *self = Frontiers::ID(id);
+                    }
+                    _ => {}
+                }
+            }
+        }
+    }
+
+    pub fn remove(&mut self, id: &ID) {
+        match self {
+            Frontiers::None => {}
+            Frontiers::ID(existing_id) => {
+                if existing_id == id {
+                    *self = Frontiers::None;
+                }
+            }
+            Frontiers::Map(map) => {
+                if map.remove(id) {
+                    match map.len() {
+                        0 => *self = Frontiers::None,
+                        1 => {
+                            let id = map.iter().next().unwrap();
+                            *self = Frontiers::ID(id);
+                        }
+                        _ => {}
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl PartialEq for Frontiers {
+    fn eq(&self, other: &Self) -> bool {
+        let len = self.len();
+        if len != other.len() {
+            return false;
+        }
+
+        match (self, other) {
+            (Frontiers::None, Frontiers::None) => true,
+            (Frontiers::ID(id1), Frontiers::ID(id2)) => id1 == id2,
+            (Frontiers::Map(map1), Frontiers::Map(map2)) => map1 == map2,
+            (Frontiers::ID(id), Frontiers::Map(map)) | (Frontiers::Map(map), Frontiers::ID(id)) => {
```

Comment:
why?

Reply:
The map may contain only one element sometimes. Maybe I should enforce the behavior of at least 2 elements for map